### PR TITLE
[Feature] Add reference SQL provenance sidecar

### PR DESCRIPTION
**Pull request by stukid**
Summary
- Add optional agent.knowledge_base.provenance sidecar storage for benchmark/evaluation metadata.
- Parse reference SQL source_context directives from SQL comments, strip them before SQL validation/storage, and write provenance rows after successful bootstrap.
- Enrich reference_sql search/detail results with source_ids/source_context_ids/source_metadata when provenance is enabled.

Validation
- uv run ruff check datus/configuration/agent_config.py datus/storage/knowledge_provenance.py datus/storage/reference_sql/sql_file_processor.py datus/storage/reference_sql/reference_sql_init.py datus/storage/reference_sql/store.py tests/unit_tests/storage/test_knowledge_provenance.py tests/unit_tests/storage/reference_sql/test_sql_file_processor.py
- uv run pytest tests/unit_tests/storage/test_knowledge_provenance.py tests/unit_tests/storage/reference_sql/test_sql_file_processor.py -q
- uv run pytest tests/unit_tests/storage/reference_sql/test_reference_sql_init.py -q

Refs #856

## Backport

- [ ] release/0.3.1
- [ ] release/0.3.2
## New Features
- Added optional knowledge provenance tracking system to store and display metadata about the source and origin of reference SQL items
- SQL files now support embedded metadata directives in comments for annotating source identifiers, context IDs, types, and additional information
- Reference SQL search results are automatically enriched with stored provenance information when the feature is enabled in configuration

[Review Change Stack](https://app.coderabbit.ai/change-stack/Datus-ai/Datus-agent/pull/925?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

## Summary by CodeRabbit

* **New Features**
  * Added a knowledge_base setting in agent config.
  * Optional file-backed knowledge-provenance sidecar to record/persist source metadata for reference SQL.
  * SQL files may include embedded provenance directives to annotate sources.
  * Reference-SQL flows (init/search/CLI) can sync provenance and report synced counts.

* **Bug Fixes / Reliability**
  * Robust handling of corrupt/missing on-disk provenance data; preserves original creation timestamps on updates.

* **Tests**
  * Expanded unit tests covering parsing, persistence, enrichment, toggles, and edge cases.

### CODE REVIEW

File: datus/configuration/agent_config.py [hunk 1]
```diff
@@ -761,6 +762,7 @@ def __init__(self, nodes: Dict[str, NodeConfig], **kwargs):
         # inject their own via :meth:`set_provider_catalog`.
         self._provider_catalog: Optional[Dict[str, Any]] = None
         self._benchmark_config_dict = kwargs.get("benchmark", {})
+        self.knowledge_base: Dict[str, Any] = kwargs.get("knowledge_base", {}) or {}
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/configuration/agent_config.py]: Fixed by resolving `agent.knowledge_base` through `_resolve_nested_value` and defaulting non-dict values to `{}`. Added AgentConfig coverage for nested env placeholders. Validation: `uv run pytest tests/unit_tests/configuration/test_agent_config.py::TestAgentConfigKnowledgeBase -q`; `uv run python ci/run-pr-tests.py upstream/main`.

File: datus/storage/knowledge_provenance.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+            key = self._row_key(normalized)
+            created_at = by_key.get(key, {}).get("created_at") or now
+            normalized["created_at"] = created_at
+            normalized["updated_at"] = now
+            by_key[key] = normalized
+            written += 1
+
+        if written:
+            self._write_rows(sorted(by_key.values(), key=self._row_key))
+        return written
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/storage/knowledge_provenance.py]: Fixed by serializing the sidecar read/merge/write cycle with a lock file around `upsert_many()` and replacement writes. Validation: `uv run pytest tests/unit_tests/storage/test_knowledge_provenance.py -q`; `uv run python ci/run-pr-tests.py upstream/main`.

File: tests/unit_tests/storage/test_knowledge_provenance.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+
+def _config(tmp_path, enabled=True):
+    return SimpleNamespace(
+        knowledge_base={"provenance": {"enabled": enabled}},
+        path_manager=SimpleNamespace(project_data_dir=tmp_path),
+    )
+
+
+@pytest.mark.ci
```
[the remark this answers is not in the document]
Comment (stukid) [reply tests/unit_tests/storage/test_knowledge_provenance.py]: Fixed by removing the redundant `@pytest.mark.ci` markers and the now-unused `pytest` import from the provenance unit tests. Validation: `uv run pytest tests/unit_tests/storage/test_knowledge_provenance.py -q`; `uv run python ci/run-pr-tests.py upstream/main`.

File: datus/storage/reference_sql/reference_sql_init.py [hunk 4]
```diff
@@ -24,6 +29,23 @@
...
+    if not rows:
+        return 0
+
+    try:
+        written = KnowledgeProvenanceStore(agent_config).upsert_many(rows)
+        logger.info("Synced %d reference SQL provenance row(s)", written)
+        return written
+    except Exception as exc:
+        logger.warning("Failed to sync reference SQL provenance sidecar: %s", exc)
+        return 0
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/storage/reference_sql/reference_sql_init.py:55]: Fixed by changing bootstrap provenance sync to replace all rows for the affected reference SQL artifact IDs. This deletes stale context/source rows when provenance is removed or changed. Validation: `uv run pytest tests/unit_tests/storage/reference_sql/test_reference_sql_init.py -q`; `uv run python ci/run-pr-tests.py upstream/main`.

File: tests/unit_tests/storage/reference_sql/test_sql_file_processor.py [hunk 5]
```diff
@@ -582,3 +583,44 @@ def test_log_multiple_entries(self, tmp_path, monkeypatch):
...
         assert "[2]" in content
+
+
+# ============================================================
+# Tests for provenance directives
+# ============================================================
+
+
+@pytest.mark.ci
+class TestSqlProvenanceDirectives:
```
[the remark this answers is not in the document]
Comment (stukid) [reply tests/unit_tests/storage/reference_sql/test_sql_file_processor.py]: Fixed by removing the redundant `@pytest.mark.ci` marker from the new unit-test class and adding an assertion that provenance directives are stripped from the stored SQL. Validation: `uv run pytest tests/unit_tests/storage/reference_sql/test_sql_file_processor.py::TestSqlProvenanceDirectives -q`; `uv run python ci/run-pr-tests.py upstream/main`.

File: datus/storage/reference_sql/sql_file_processor.py [hunk 6]
```diff
@@ -234,31 +311,39 @@ def process_sql_items(items: List[Dict[str, Any]]) -> Tuple[List[Dict[str, Any]]
         comment = item.get("comment") or ""
         filepath = item.get("filepath") or ""
         line_number = item.get("line_number", 1)
+        provenance = extract_sql_provenance(sql)
+        sql_for_processing = strip_sql_provenance_directives(sql) if provenance else sql
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/storage/reference_sql/sql_file_processor.py:316]: Fixed. The processor now strips provenance directives unconditionally before SQL type detection and validation, even when the directive values are empty. I also added a regression test for an empty source_context_id directive. Validation: uv run pytest tests/unit_tests/storage/reference_sql/test_sql_file_processor.py -q; uv run ruff check datus/storage/reference_sql/sql_file_processor.py tests/unit_tests/storage/reference_sql/test_sql_file_processor.py.
